File: clients/ts/src/baml_lib/deserializer/diagnostics.ts
type Scope = string[];

interface DeserializerErrorInterface {
    scope: Scope;
    message: string;
    toWarning(): DeserializerWarning;
}

class DeserializerError implements DeserializerErrorInterface {
    
    constructor(private _scope: Scope, private _message: string) {
    }

    get message(): string {
        return this._message;
    }

    get scope(): Scope {
        return this._scope;
    }

    toString(): string {
        return this.scope.length === 0 ? `Error: ${this._message}` : `Error in ${this.scope.join('.')}: ${this._message}`;
    }

    toWarning(): DeserializerWarning {
        return new DeserializerWarning(this.scope, this._message);
    }
}

class DeserializerWarning {
    constructor(private _scope: Scope, private _message: string) {
    }


    get scope(): Scope {
        return this._scope;
    }

    toString(): string {
        return this.scope.length === 0 ? `Warning: ${this._message}` : `Warning in ${this.scope.join('.')}: ${this._message}`;
    }
}

class DeserializerException extends Error {
    private errors: DeserializerError[];
    private warnings: DeserializerWarning[];
    private rawString: string;

    constructor(errors: DeserializerError[], warnings: DeserializerWarning[], rawString: string) {
        super(`Failed to Deserialize: (${errors.length} errors) (${warnings.length} warnings)`);
        this.errors = errors;
        this.warnings = warnings;
        this.rawString = rawString;
    }

    toString(): string {
        let output: string[] = [this.message];
        let items = [...this.errors, ...this.warnings];
        items.sort((a, b) => b.scope.length - a.scope.length);
        items.forEach(i => {
            output.push('------');
            output.push(i.toString());
        });
        output.push('------');
        output.push('Raw:');
        output.push(this.rawString);
        return output.join('\n');
    }
}

class Diagnostics {
    private errors: DeserializerError[] = [];
    private warnings: DeserializerWarning[] = [];
    private rawString: string;
    private scope: Scope = [];
    private scopeErrors: { [key: string]: DeserializerError[] } = {};

    constructor(rawString: string) {
        this.rawString = rawString;
    }

    pushScope(scope: string): void {
        this.scope.push(scope);
        this.scopeErrors[this.scope.join('.')] = [];
    }

    popScope(errorsAsWarnings: boolean): void {
        let prevScopeKey = this.scope.join('.');
        this.scope.pop();
        // If there are any errors, convert them to warnings.
        this.scopeErrors[prevScopeKey]?.forEach(error => {
            if (errorsAsWarnings) {
                this.pushWarning(error.toWarning());
            } else {
                this.pushError(error);
            }
        });
    }

    private pushError(error: DeserializerError): void {
        let key = this.scope.join('.');
        if (key) {
            this.scopeErrors[key]?.push(error);
        } else {
            this.errors.push(error);
        }
    }

    private pushWarning(warning: DeserializerWarning): void {
        this.warnings.push(warning);
    }

    toException(): void {
        if (this.errors.length > 0) {
            throw new DeserializerException(this.errors, this.warnings, this.rawString);
        }
    }

    pushUnknownWarning(message: string): void {
        this.pushWarning(new DeserializerWarning(this.scope, message));
    }

    pushEnumError(enumName: string, value: any, expected: string[]): void {
        this.pushError(new DeserializerError(this.scope, `Failed to parse \`${value}\` as \`${enumName}\`. Expected one of: ${expected.join(', ')}`));
    }

    pushUnknownError(message: string): void {
        this.pushError(new DeserializerError(this.scope, message));
    }
}

export { DeserializerError, DeserializerWarning, DeserializerException, Diagnostics };

File: clients/python/baml_lib/_impl/deserializer/diagnostics.py
import typing


class DeserializerException(Exception):
    def __init__(
        self,
        errors: typing.List["DeserializerError"],
        warnings: typing.List["DeserializerWarning"],
        raw_string: str,
    ) -> None:
        super().__init__(errors)
        self.__items = errors + warnings
        self.__items.sort(key=lambda x: x.scope, reverse=True)
        self.__num_errors = len(errors)
        self.__num_warnings = len(warnings)
        self.__raw_string = raw_string

    def __str__(self) -> str:
        output = [
            f"Failed to Deserialize: ({self.__num_errors} errors) ({self.__num_warnings} warnings)"
        ]
        for i in self.__items:
            output.append("------")
            output.append(str(i))
        output.append("------")
        output.append("Raw:")
        output.append(self.__raw_string)
        return "\n".join(output)


class Diagnostics:
    __scope_errors: typing.Dict[str, typing.List["DeserializerError"]]
    __errors: typing.List["DeserializerError"]
    __warnings: typing.List["DeserializerWarning"]
    __scope: typing.List[str]

    def __init__(self, raw_string: str) -> None:
        self.__errors = []
        self.__warnings = []
        self.__raw_string = raw_string
        self.__scope = []
        self.__scope_errors = {}

    def push_scope(self, scope: str) -> None:
        self.__scope.append(scope)
        self.__scope_errors[".".join(self.__scope)] = []

    def pop_scope(self, errors_as_warnings: bool) -> None:
        prev_scope_key = ".".join(self.__scope)
        self.__scope.pop()
        # If there are any errors, convert them to warnings.
        for error in self.__scope_errors.get(prev_scope_key, []):
            if errors_as_warnings:
                self.__push_warning(error.to_warning())
            else:
                self.__push_error(error)

    def __push_error(self, error: "DeserializerError") -> None:
        if len(self.__scope) > 0:
            key = ".".join(self.__scope)
            if key in self.__scope_errors:
                self.__scope_errors[key].append(error)
        else:
            self.__errors.append(error)

    def __push_warning(self, warning: "DeserializerWarning") -> None:
        self.__warnings.append(warning)

    def to_exception(self) -> None:
        """
        This method raises a DeserializerException if there are any errors in the diagnostics.
        """
        if len(self.__errors) > 0:
            raise DeserializerException(
                self.__errors, self.__warnings, self.__raw_string
            )

    def push_unkown_warning(self, message: str) -> None:
        self.__push_warning(DeserializerWarning(self.__scope, message))

    def push_missing_keys_error(self, keys: typing.List[str]) -> None:
        self.__push_error(
            DeserializerError(self.__scope, f"Missing keys: {', '.join(keys)}")
        )

    def push_enum_error(
        self, enum_name: str, value: typing.Any, expected: typing.List[str]
    ) -> None:
        self.__push_error(
            DeserializerError(
                self.__scope,
                f'Failed to parse `{value}` as `{enum_name}`. Expected one of: {", ".join(expected)}',
            )
        )

    def push_unknown_error(self, message: str) -> None:
        self.__push_error(DeserializerError(self.__scope, message))


class DeserializerError:
    def __init__(self, scope: typing.List[str], message: str):
        self.__message = message
        self.__scope = list(scope)

    @property
    def scope(self) -> typing.List[str]:
        return self.__scope

    def __str__(self) -> str:
        if len(self.__scope) == 0:
            return f"Error: {self.__message}"
        return f"Error in {'.'.join(self.__scope)}: {self.__message}"

    def __repr__(self) -> str:
        if len(self.__scope) == 0:
            return f"Error: {self.__message}"
        return f"Error in {'.'.join(self.__scope)}: {self.__message}"

    def to_warning(self) -> "DeserializerWarning":
        return DeserializerWarning(self.__scope, self.__message)


class DeserializerWarning:
    def __init__(self, scope: typing.List[str], message: str):
        self.__message = message
        self.__scope = list(scope)

    @property
    def scope(self) -> typing.List[str]:
        return self.__scope

    def __str__(self) -> str:
        if len(self.__scope) == 0:
            return f"Warning: {self.__message}"
        return f"Warning in {'.'.join(self.__scope)}: {self.__message}"

    def __repr__(self) -> str:
        if len(self.__scope) == 0:
            return f"Warning: {self.__message}"
        return f"Warning in {'.'.join(self.__scope)}: {self.__message}"
